diff --git a/changes/1004.bugfix.rst b/changes/1004.bugfix.rst
new file mode 100644
index 000000000..f881c8ce3
--- /dev/null
+++ b/changes/1004.bugfix.rst
@@ -0,0 +1,1 @@
+Web project configuration has been updated to reflect recent changes to PyScript.
diff --git a/src/briefcase/platforms/web/static.py b/src/briefcase/platforms/web/static.py
index 8aaef4fc0..93576ce9a 100644
--- a/src/briefcase/platforms/web/static.py
+++ b/src/briefcase/platforms/web/static.py
@@ -170,7 +170,8 @@ def build_app(self, app: AppConfig, **kwargs):
                     "name": app.formal_name,
                     "description": app.description,
                     "version": app.version,
-                    "autoclose_loader": True,
+                    "splashscreen": {"autoclose": True},
+                    "terminal": False,
                     # Ensure that we're using Unix path separators, as the content
                     # will be parsed by pyscript in the browser.
                     "packages": [
