diff --git a/tests/platforms/web/static/test_build.py b/tests/platforms/web/static/test_build.py
index 910e08c77..b08d93c9b 100644
--- a/tests/platforms/web/static/test_build.py
+++ b/tests/platforms/web/static/test_build.py
@@ -116,7 +116,8 @@ def mock_run(*args, **kwargs):
             "name": "First App",
             "description": "The first simple app \\ demonstration",
             "version": "0.0.1",
-            "autoclose_loader": True,
+            "splashscreen": {"autoclose": True},
+            "terminal": False,
             "packages": [
                 "/static/wheels/dependency-1.2.3-py3-none-any.whl",
                 "/static/wheels/first_app-1.2.3-py3-none-any.whl",
@@ -288,7 +289,8 @@ def mock_run(*args, **kwargs):
             "name": "First App",
             "description": "The first simple app \\ demonstration",
             "version": "0.0.1",
-            "autoclose_loader": True,
+            "splashscreen": {"autoclose": True},
+            "terminal": False,
             "packages": [
                 "/static/wheels/first_app-1.2.3-py3-none-any.whl",
             ],
