diff --git a/CTFd/admin/__init__.py b/CTFd/admin/__init__.py
index 0a5410a251..fe40cc82fd 100644
--- a/CTFd/admin/__init__.py
+++ b/CTFd/admin/__init__.py
@@ -43,6 +43,7 @@
 )
 from CTFd.utils import config as ctf_config
 from CTFd.utils import get_config, set_config
+from CTFd.utils.csv import load_challenges_csv, load_teams_csv, load_users_csv
 from CTFd.utils.decorators import admins_only
 from CTFd.utils.exports import export_ctf as export_ctf_util
 from CTFd.utils.exports import import_ctf as import_ctf_util
@@ -116,6 +117,33 @@ def export_ctf():
     )
 
 
+@admin.route("/admin/import/csv", methods=["POST"])
+@admins_only
+def import_csv():
+    csv_type = request.form["csv_type"]
+    # Try really hard to load data in properly no matter what nonsense Excel gave you
+    raw = request.files["csv_file"].stream.read()
+    try:
+        csvdata = raw.decode("utf-8-sig")
+    except UnicodeDecodeError:
+        try:
+            csvdata = raw.decode("cp1252")
+        except UnicodeDecodeError:
+            csvdata = raw.decode("latin-1")
+    csvfile = StringIO(csvdata)
+
+    loaders = {
+        "challenges": load_challenges_csv,
+        "users": load_users_csv,
+        "teams": load_teams_csv,
+    }
+
+    loader = loaders[csv_type]
+    reader = csv.DictReader(csvfile)
+    loader(reader)
+    return redirect(url_for("admin.config"))
+
+
 @admin.route("/admin/export/csv")
 @admins_only
 def export_csv():
diff --git a/CTFd/forms/config.py b/CTFd/forms/config.py
index 61e0ac4916..a9c1929a5f 100644
--- a/CTFd/forms/config.py
+++ b/CTFd/forms/config.py
@@ -1,4 +1,4 @@
-from wtforms import BooleanField, SelectField, StringField, TextAreaField
+from wtforms import BooleanField, FileField, SelectField, StringField, TextAreaField
 from wtforms.fields.html5 import IntegerField, URLField
 from wtforms.widgets.html5 import NumberInput
 
@@ -83,6 +83,15 @@ class ExportCSVForm(BaseForm):
     submit = SubmitField("Download CSV")
 
 
+class ImportCSVForm(BaseForm):
+    csv_type = SelectField(
+        "CSV Type",
+        choices=[("users", "Users"), ("teams", "Teams"), ("challenges", "Challenges")],
+        description="Type of CSV data",
+    )
+    csv_file = FileField("CSV File", description="CSV file contents")
+
+
 class LegalSettingsForm(BaseForm):
     tos_url = URLField(
         "Terms of Service URL",
diff --git a/CTFd/themes/admin/templates/configs/backup.html b/CTFd/themes/admin/templates/configs/backup.html
index 743d9a85ec..c4221223e4 100644
--- a/CTFd/themes/admin/templates/configs/backup.html
+++ b/CTFd/themes/admin/templates/configs/backup.html
@@ -10,6 +10,9 @@
 		<li class="nav-item">
 			<a class="nav-link " href="#save-csv" aria-controls="save-csv" role="tab" data-toggle="tab">Download CSV</a>
 		</li>
+		<li class="nav-item">
+			<a class="nav-link " href="#import-csv" aria-controls="import-csv" role="tab" data-toggle="tab">Import CSV</a>
+		</li>
 	</ul>
 	<div class="tab-content">
 		<div role="tabpanel" class="tab-pane active" id="export-ctf">
@@ -49,5 +52,27 @@
 			</form>
 			{% endwith %}
 		</div>
+		<div role="tabpanel" class="tab-pane" id="import-csv">
+			{% with form = Forms.config.ImportCSVForm() %}
+			<form method="POST" action="{{ url_for('admin.import_csv') }}" enctype="multipart/form-data">
+				<div class="form-group">
+					<b>{{ form.csv_type.label }}</b>
+					{{ form.csv_type(class="form-control custom-select") }}
+					<small class="form-text text-muted">
+						{{ form.csv_type.description }}
+					</small>
+				</div>
+				<div class="form-group">
+					<b>{{ form.csv_file.label }}</b>
+					{{ form.csv_file(class="form-control-file") }}
+					<small class="form-text text-muted">
+						{{ form.csv_file.description }}
+					</small>
+				</div>
+				{{ form.nonce() }}
+				<input type="submit" class="btn btn-warning" value="Import CSV">
+			</form>
+			{% endwith %}
+		</div>
 	</div>
 </div>
\ No newline at end of file
diff --git a/CTFd/utils/csv/__init__.py b/CTFd/utils/csv/__init__.py
new file mode 100644
index 0000000000..ac6cc1ffd0
--- /dev/null
+++ b/CTFd/utils/csv/__init__.py
@@ -0,0 +1,59 @@
+import json
+
+from CTFd.models import Flags, Hints, Tags, Teams, Users, db
+from CTFd.plugins.challenges import get_chal_class
+
+
+def load_users_csv(dict_reader):
+    for line in dict_reader:
+        result = Users(**line)
+        db.session.add(result)
+        db.session.commit()
+    return True
+
+
+def load_teams_csv(dict_reader):
+    for line in dict_reader:
+        result = Teams(**line)
+        db.session.add(result)
+        db.session.commit()
+    return True
+
+
+def load_challenges_csv(dict_reader):
+    for line in dict_reader:
+        flags = line.pop("flags", None)
+        tags = line.pop("tags", None)
+        hints = line.pop("hints", None)
+        challenge_type = line.pop("type", "standard")
+
+        # Load in custome type_data
+        type_data = json.loads(line.pop("type_data", "{}"))
+        line.update(type_data)
+
+        ChallengeClass = get_chal_class(challenge_type)
+        challenge = ChallengeClass.challenge_model(**line)
+        db.session.add(challenge)
+        db.session.commit()
+
+        if flags:
+            flags = [flag.strip() for flag in flags.split(",")]
+            for flag in flags:
+                f = Flags(type="static", challenge_id=challenge.id, content=flag,)
+                db.session.add(f)
+                db.session.commit()
+
+        if tags:
+            tags = [tag.strip() for tag in tags.split(",")]
+            for tag in tags:
+                t = Tags(challenge_id=challenge.id, value=tag,)
+                db.session.add(t)
+                db.session.commit()
+
+        if hints:
+            hints = [hint.strip() for hint in hints.split(",")]
+            for hint in hints:
+                h = Hints(challenge_id=challenge.id, content=hint,)
+                db.session.add(h)
+                db.session.commit()
+    return True
