diff --git a/tests/test_plugin_utils.py b/tests/test_plugin_utils.py
index c4c958878b..24403fb147 100644
--- a/tests/test_plugin_utils.py
+++ b/tests/test_plugin_utils.py
@@ -14,7 +14,13 @@
     register_plugin_script,
     register_user_page_menu_bar,
 )
-from tests.helpers import create_ctfd, destroy_ctfd, login_as_user, setup_ctfd
+from tests.helpers import (
+    create_ctfd,
+    destroy_ctfd,
+    gen_challenge,
+    login_as_user,
+    setup_ctfd,
+)
 
 
 def test_register_plugin_asset():
@@ -204,3 +210,17 @@ def bypass_csrf_protection_test_route():
             assert r.status_code == 200
             assert output == "Success"
     destroy_ctfd(app)
+
+
+def test_challenges_model_access_plugin_class():
+    """
+    Test that the Challenges model can access its plugin class
+    """
+    app = create_ctfd()
+
+    with app.app_context():
+        from CTFd.plugins.challenges import get_chal_class
+
+        chal = gen_challenge(app.db)
+        assert chal.plugin_class == get_chal_class("standard")
+    destroy_ctfd(app)
