diff --git a/poetry/packages/dependency.py b/poetry/packages/dependency.py
index 31a360ebdbf..74d8f92c1d4 100644
--- a/poetry/packages/dependency.py
+++ b/poetry/packages/dependency.py
@@ -200,6 +200,9 @@ def to_pep_508(self, with_extras=True):  # type: (bool) -> str
             )
 
         if markers:
+            if self.is_vcs():
+                requirement += " "
+
             if len(markers) > 1:
                 markers = ["({})".format(m) for m in markers]
                 requirement += "; {}".format(" and ".join(markers))
