diff --git a/tests/packages/test_vcs_dependency.py b/tests/packages/test_vcs_dependency.py
index 18e3e978d80..9872a443a7c 100644
--- a/tests/packages/test_vcs_dependency.py
+++ b/tests/packages/test_vcs_dependency.py
@@ -20,3 +20,36 @@ def test_to_pep_508_with_extras():
     expected = "poetry[foo] @ git+https://github.com/sdispater/poetry.git@master"
 
     assert expected == dependency.to_pep_508()
+
+
+def test_to_pep_508_in_extras():
+    dependency = VCSDependency(
+        "poetry", "git", "https://github.com/sdispater/poetry.git"
+    )
+    dependency.in_extras.append("foo")
+
+    expected = (
+        'poetry @ git+https://github.com/sdispater/poetry.git@master ; extra == "foo"'
+    )
+    assert expected == dependency.to_pep_508()
+
+    dependency = VCSDependency(
+        "poetry", "git", "https://github.com/sdispater/poetry.git"
+    )
+    dependency.in_extras.append("foo")
+    dependency.extras.append("bar")
+
+    expected = 'poetry[bar] @ git+https://github.com/sdispater/poetry.git@master ; extra == "foo"'
+
+    assert expected == dependency.to_pep_508()
+
+    dependency = VCSDependency(
+        "poetry", "git", "https://github.com/sdispater/poetry.git", "b;ar;"
+    )
+    dependency.in_extras.append("foo;")
+
+    expected = (
+        'poetry @ git+https://github.com/sdispater/poetry.git@b;ar; ; extra == "foo;"'
+    )
+
+    assert expected == dependency.to_pep_508()
