diff --git a/CHANGELOG.md b/CHANGELOG.md
index a0bae8d55..1915a9d3c 100644
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,12 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## Unreleased
+
+### Fixed
+
+- Fixed `Table` columns not highlighting when added by `add_row` https://github.com/Textualize/rich/issues/3517
+
 ## [13.9.1] - 2024-10-01
 
 ### Fixed
diff --git a/rich/table.py b/rich/table.py
index 10af2e6f4..7b1fa9e4b 100644
--- a/rich/table.py
+++ b/rich/table.py
@@ -451,7 +451,7 @@ def add_cell(column: Column, renderable: "RenderableType") -> None:
             ]
         for index, renderable in enumerate(cell_renderables):
             if index == len(columns):
-                column = Column(_index=index)
+                column = Column(_index=index, highlight=self.highlight)
                 for _ in self.rows:
                     add_cell(column, Text(""))
                 self.columns.append(column)
