diff --git a/CHANGELOG.rst b/CHANGELOG.rst
index a68baa71ff..c5927c4073 100644
--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -13,6 +13,7 @@ This project adheres to `Semantic Versioning <https://semver.org/>`_.
 * Added ``--ciphers`` to allow configuring OpenSSL ciphers (`#870`_).
 * Added support for ``$XDG_CONFIG_HOME`` (`#920`_).
 * Fixed built-in plugins-related circular imports (`#925`_).
+* Fixed custom content types for each multipart uploaded file (`#668`_).
 
 
 `2.1.0`_ (2020-04-18)
@@ -441,3 +442,4 @@ This project adheres to `Semantic Versioning <https://semver.org/>`_.
 .. _#895: https://github.com/jakubroztocil/httpie/issues/895
 .. _#920: https://github.com/jakubroztocil/httpie/issues/920
 .. _#925: https://github.com/jakubroztocil/httpie/issues/925
+.. _#668: https://github.com/jakubroztocil/httpie/issues/668
diff --git a/README.rst b/README.rst
index b287c9e2c5..2168855200 100644
--- a/README.rst
+++ b/README.rst
@@ -683,6 +683,13 @@ submitted:
 Note that ``@`` is used to simulate a file upload form field, whereas
 ``=@`` just embeds the file content as a regular text field value.
 
+When uploading files, their content type is inferred from the file name. You can manually
+override the inferred content type:
+
+.. code-block:: bash
+
+   $ http -f POST httpbin.org/post name='John Smith' cv@'~/files/data.bin;type=application/pdf'
+
 
 HTTP headers
 ============
diff --git a/httpie/cli/constants.py b/httpie/cli/constants.py
index 27d0eb259e..ea8036fe24 100644
--- a/httpie/cli/constants.py
+++ b/httpie/cli/constants.py
@@ -24,6 +24,7 @@
 SEPARATOR_DATA_STRING = '='
 SEPARATOR_DATA_RAW_JSON = ':='
 SEPARATOR_FILE_UPLOAD = '@'
+SEPARATOR_FILE_UPLOAD_TYPE = ';type='  # in already parsed file upload path only
 SEPARATOR_DATA_EMBED_FILE_CONTENTS = '=@'
 SEPARATOR_DATA_EMBED_RAW_JSON_FILE = ':=@'
 SEPARATOR_QUERY_PARAM = '=='
diff --git a/httpie/cli/definition.py b/httpie/cli/definition.py
index 2d429260c8..93fef94e86 100644
--- a/httpie/cli/definition.py
+++ b/httpie/cli/definition.py
@@ -113,6 +113,7 @@
     '@' Form file fields (only with --form, -f):
 
         cs@~/Documents/CV.pdf
+        cv@'~/Documents/CV.pdf;type=application/pdf'
 
     '=@' A data field like '=', but takes a file path and embeds its content:
 
diff --git a/httpie/cli/requestitems.py b/httpie/cli/requestitems.py
index ec8e6c1ef1..a812d65174 100644
--- a/httpie/cli/requestitems.py
+++ b/httpie/cli/requestitems.py
@@ -6,7 +6,8 @@
 from httpie.cli.constants import (
     SEPARATOR_DATA_EMBED_FILE_CONTENTS, SEPARATOR_DATA_EMBED_RAW_JSON_FILE,
     SEPARATOR_DATA_RAW_JSON, SEPARATOR_DATA_STRING, SEPARATOR_FILE_UPLOAD,
-    SEPARATOR_HEADER, SEPARATOR_HEADER_EMPTY, SEPARATOR_QUERY_PARAM,
+    SEPARATOR_FILE_UPLOAD_TYPE, SEPARATOR_HEADER, SEPARATOR_HEADER_EMPTY,
+    SEPARATOR_QUERY_PARAM,
 )
 from httpie.cli.dicts import (
     RequestDataDict, RequestFilesDict, RequestHeadersDict, RequestJSONDataDict,
@@ -95,7 +96,10 @@ def process_query_param_arg(arg: KeyValueArg) -> str:
 
 
 def process_file_upload_arg(arg: KeyValueArg) -> Tuple[str, IO, str]:
-    filename = arg.value
+    parts = arg.value.split(SEPARATOR_FILE_UPLOAD_TYPE)
+    filename = parts[0]
+    mime_type = parts[1] if len(parts) > 1 else None
+
     try:
         with open(os.path.expanduser(filename), 'rb') as f:
             contents = f.read()
@@ -104,7 +108,7 @@ def process_file_upload_arg(arg: KeyValueArg) -> Tuple[str, IO, str]:
     return (
         os.path.basename(filename),
         BytesIO(contents),
-        get_content_type(filename),
+        mime_type or get_content_type(filename),
     )
 
 
