diff --git a/tests/api/v1/test_config.py b/tests/api/v1/test_config.py
index 57b27ce3fe..5c7f2c0cb5 100644
--- a/tests/api/v1/test_config.py
+++ b/tests/api/v1/test_config.py
@@ -105,21 +105,73 @@ def test_api_config_delete_admin():
     destroy_ctfd(app)
 
 
-def test_long_values():
-    """Can a config value that is bigger than 64,000 be accepted"""
+def test_config_value_types():
+    """Test that we properly receive values according to schema"""
     app = create_ctfd()
     with app.app_context():
         with login_as_user(app, "admin") as admin:
-            long_text = "a" * 65000
+            # Test new configs error out if too long
+            long_text = "a" * 65536
             r = admin.post(
-                "/api/v1/configs", json={"key": "ctf_footer", "value": long_text}
+                "/api/v1/configs", json={"key": "new_ctf_config", "value": long_text}
             )
             data = r.get_json()
-            assert data["errors"]["value"][0] == "ctf_footer config is too long"
+            assert data["errors"]["value"][0] == "new_ctf_config config is too long"
+            assert r.status_code == 400
+            r = admin.post(
+                "/api/v1/configs", json={"key": "new_ctf_config", "value": "test"}
+            )
+            assert r.status_code == 200
+            assert get_config("new_ctf_config") == "test"
 
-            r = admin.patch("/api/v1/configs", json={"ctf_theme": long_text})
+            # Test strings too long error out
+            r = admin.patch("/api/v1/configs", json={"ctf_footer": long_text})
             data = r.get_json()
-            assert data["errors"]["value"][0] == "ctf_theme config is too long"
+            assert data["errors"]["value"][0] == "ctf_footer config is too long"
             assert r.status_code == 400
 
+            # Test regular length strings
+            r = admin.patch(
+                "/api/v1/configs", json={"ctf_footer": "// regular length string"},
+            )
+            assert r.status_code == 200
+            assert get_config("ctf_footer") == "// regular length string"
+
+            # Test booleans can be received
+            r = admin.patch("/api/v1/configs", json={"view_after_ctf": True})
+            assert r.status_code == 200
+            assert bool(get_config("view_after_ctf")) == True
+
+            # Test None can be received
+            assert get_config("mail_username") is None
+            r = admin.patch("/api/v1/configs", json={"mail_username": "testusername"})
+            assert r.status_code == 200
+            assert get_config("mail_username") == "testusername"
+            r = admin.patch("/api/v1/configs", json={"mail_username": None})
+            assert r.status_code == 200
+            assert get_config("mail_username") is None
+
+            # Test integers can be received
+            r = admin.patch("/api/v1/configs", json={"mail_port": 12345})
+            assert r.status_code == 200
+            assert get_config("mail_port") == 12345
+
+            # Test specific config key
+            r = admin.patch(
+                "/api/v1/configs/long_config_test", json={"value": long_text}
+            )
+            data = r.get_json()
+            assert data["errors"]["value"][0] == "long_config_test config is too long"
+            assert r.status_code == 400
+            assert get_config("long_config_test") is None
+            r = admin.patch(
+                "/api/v1/configs/config_test", json={"value": "config_value_test"}
+            )
+            assert r.status_code == 200
+            assert get_config("config_test") == "config_value_test"
+            r = admin.patch(
+                "/api/v1/configs/mail_username", json={"value": "testusername"}
+            )
+            assert r.status_code == 200
+            assert get_config("mail_username") == "testusername"
     destroy_ctfd(app)
